test/generators/gen_link/gen_link.ml

```diff
@@ -0,0 +1,214 @@
+            [
+              Atom "run";
+              Atom "ocamlc";
+              Atom "-c";
+              Atom "-bin-annot";
+              Atom "-o";
+              Atom "%{target}";
+              Atom "%{deps}";
+            ];
+        ];
+    ]
+
+let odoc_target_rule basename ext =
+  List
+    [
+      Atom "rule";
+      List [ Atom "target"; Atom (basename ^ ".odoc") ];
+      List [ Atom "deps"; Atom (basename ^ ext) ];
+      List
+        [
+          Atom "action";
+          List
+            [
+              Atom "run";
+              Atom "odoc";
+              Atom "compile";
+              Atom "--pkg";
+              Atom "test";
+              Atom "-o";
+              Atom "%{target}";
+              Atom "%{deps}";
+            ];
+        ];
+    ]
+
+let odocl_target_rule basename =
+  List
+    [
+      Atom "rule";
+      List [ Atom "target"; Atom (basename ^ ".odocl") ];
+      List [ Atom "deps"; Atom (basename ^ ".odoc") ];
+      List
+        [
+          Atom "action";
+          List
+            [
+              Atom "run";
+              Atom "odoc";
+              Atom "link";
+              Atom "-o";
+              Atom "%{target}";
+              Atom "%{deps}";
+            ];
+        ];
+    ]
+
+let file_rule basename source_file ext =
+  let open Printf in
+  printf "%s\n\n" (to_string_hum (cu_target_rule basename source_file ext));
+  printf "%s\n\n" (to_string_hum (odoc_target_rule basename ext));
+  printf "%s\n\n" (to_string_hum (odocl_target_rule basename))
+
+let mld_odoc_target_rule basename source_file =
+  List
+    [
+      Atom "rule";
+      List [ Atom "target"; Atom ("page-" ^ basename ^ ".odoc") ];
+      List [ Atom "deps"; Atom source_file ];
+      List
+        [
+          Atom "action";
+          List
+            [
+              Atom "run";
+              Atom "odoc";
+              Atom "compile";
+              Atom "--pkg";
+              Atom "test";
+              Atom "-o";
+              Atom "%{target}";
+              Atom "%{deps}";
+            ];
+        ];
+    ]
+
+let mld_odocl_target_rule basename =
+  List
+    [
+      Atom "rule";
+      List [ Atom "target"; Atom ("page-" ^ basename ^ ".odocl") ];
+      List [ Atom "deps"; Atom ("page-" ^ basename ^ ".odoc") ];
+      List
+        [
+          Atom "action";
+          List
+            [
+              Atom "run";
+              Atom "odoc";
+              Atom "link";
+              Atom "-o";
+              Atom "%{target}";
+              Atom "%{deps}";
+            ];
+        ];
+    ]
+
+let mld_file_rule basename source_file =
+  let open Printf in
+  printf "%s\n\n" (to_string_hum (mld_odoc_target_rule basename source_file));
+  printf "%s\n\n" (to_string_hum (mld_odocl_target_rule basename))
+
+let die s =
+  prerr_endline s;
+  exit 1
+
+let path () f = Filename.quote (Fpath.to_string f)
+
+let ext' () f = Filename.quote (Fpath.get_ext f)
+
+let gen_rule_for_source_file source_file f =
```

Comment:
This function shouldn't take the full path as a string and the file name as an `Fpath.t`. It should only the full path as an `Fpath.t`. `basename` below should be `Fpath.rem_ext (Fpath.basename path)`.

Reply:
Good catch

Reply:
Done!